src/CP/valueselection/learning/environment.jl

```diff
@@ -0,0 +1,43 @@
+abstract type AbstractCPEnv{TS <: AbstractTrajectoryState} <: AbstractEnv end
+
+struct CPEnv{TS} <: AbstractCPEnv{TS}
+    reward::Float32
+    terminal::Bool
+    state::TS
+    actions_index::Vector{Int}
+    legal_actions::Vector{Int}
+    legal_actions_mask::Vector{Bool}
+end
+
+RLBase.action_space(env::CPEnv) = env.actions_index
+RLBase.legal_action_space(env::CPEnv) = env.legal_actions
+RLBase.legal_action_space_mask(env::CPEnv) = env.legal_actions_mask
+RLBase.reward(env::CPEnv) = env.reward
+RLBase.is_terminated(env::CPEnv) = env.terminal
+RLBase.state(env::CPEnv) = env.state
+RLBase.ActionStyle(::CPEnv) = FULL_ACTION_SET
+RLBase.state_space(::SeaPearl.CPEnv, ::Observation{Any}, ::DefaultPlayer) = nothing
+
+struct unmaskedCPEnv{TS} <: AbstractCPEnv{TS}
+    reward::Float32
+    terminal::Bool
+    state::TS
+    actions_index::Array{Int64, 1}
+end
```

Comment:
I haven't tackled this question yet, as I figured it might be a bit out of topic of the gpu/optimization refactoring. But I agree we should discuss on the usefulness of this sutrcture.